code/iaas/service-discovery/server/src/main/java/io/cattle/platform/servicediscovery/service/impl/ServiceDiscoveryServiceImpl.java

```diff
@@ -381,9 +386,104 @@ public long getServiceNetworkId(Service service) {
+            // service, or destroyed the instance)
+            // we always try to remove the instances having names:
+            // a) that don't qualify for <environment>_<service> pattern
+            // b) have higher order in name (<environment>_<service>_4 should be removed prior to
+            // <environment>_<service>_3)
+            int currentCount = originalScale - requestedCount;
+            // a)
+            for (Instance serviceInstance : serviceInstances) {
+                if (currentCount < originalScale) {
+                    if (!isServiceGeneratedName(service, serviceInstance)) {
+                        if (serviceInstance != null) {
+                            serviceNames.add(serviceInstance.getName());
+                            currentCount++;
+                        }
+                    }
+                } else {
+                    break;
+                }
+            }
+            // b)
+            for (Integer usedId : usedIds) {
+                if (currentCount < originalScale) {
+                    String instanceName = generateServiceInstanceName(service, usedId);
+                    serviceNames.add(instanceName);
+                    currentCount++;
+                } else {
+                    break;
+                }
+            }
+        }
+
+        return serviceNames;
+    }
+
+    private boolean isServiceGeneratedName(Service service, Instance serviceInstance) {
+        Environment env = objectManager.findOne(Environment.class, ENVIRONMENT.ID, service.getEnvironmentId());
+        return serviceInstance.getName().startsWith(String.format("%s_%s", env.getName(), service.getName()));
+    }
+
+    @Override
+    public String generateServiceInstanceName(Service service, int finalOrder) {
+        Environment env = objectManager.findOne(Environment.class, ENVIRONMENT.ID, service.getEnvironmentId());
+        String name = String.format("%s_%s_%d", env.getName(), service.getName(), finalOrder);
+        return name;
+    }
+
+    private List<Integer> getServiceInstanceUsedOrderIds(Service service) {
         Environment env = objectManager.findOne(Environment.class, ENVIRONMENT.ID, service.getEnvironmentId());
-        return String.format("%s_%s_%d", env.getName(), service.getName(), order + 1);
+        // get all existing instances to check if the name is in use by the instance of the same service
+        List<Integer> usedIds = new ArrayList<>();
+        // list all the instances
+        List<? extends ServiceExposeMap> instanceServiceMaps = exposeMapDao.getNonRemovedServiceInstanceMap(service
+                .getId());
+        
+        for (ServiceExposeMap instanceServiceMap : instanceServiceMaps) {
+            Instance instance = objectManager.loadResource(Instance.class, instanceServiceMap.getInstanceId());
+            if (isServiceGeneratedName(service, instance)) {
```

Comment:
Shouldn't get affected by the service renaming as update/activate operations are guarded by States. If the service is being activated/updated, another service update can't be executed.

But you are right - its not guarded against environment update. We should def address it in a subsequent PR